driver-core/src/main/com/mongodb/internal/connection/AsynchronousSocketChannelStream.java

```diff
@@ -77,7 +90,13 @@ private void initializeSocketChannel(final AsyncCompletionHandler<Void> handler,
             SocketAddress socketAddress = socketAddressQueue.poll();
 
             try {
-                AsynchronousSocketChannel attemptConnectionChannel = AsynchronousSocketChannel.open();
+                AsynchronousSocketChannel attemptConnectionChannel;
+                if (executorService != null) {
+                    AsynchronousChannelGroup group = AsynchronousChannelGroup.withThreadPool(executorService);
+                    attemptConnectionChannel = AsynchronousSocketChannel.open(group);
```

Comment:
> we should at least `shutdown`/`shutdownNow` our group when `executorService` is being `shutdown`/`shutdownNow`.

@katcharov correctly pointed out that this is technically impossible: we can't know when the application-supplied executor is being shut down by the application. This, together with the API documentation quote from the previous comment means that we should document that an application must not shut the `ExecutorService` before `close`ing `MongoClient`. We should also document that the `ExecutorService` _"should support [direct handoff or unbounded queuing of submitted tasks](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/util/concurrent/ThreadPoolExecutor.html), and the thread that invokes the [execute](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/util/concurrent/Executor.html#execute(java.lang.Runnable)) method should never invoke the task directly"_ (this can be done by linking the [`AsynchronousChannelGroup.withThreadPool`](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/nio/channels/AsynchronousChannelGroup.html#withThreadPool(java.util.concurrent.ExecutorService)) method).

We still should not shut the application-supplied `ExecutorService` down when we close our client and shut the `AsynchronousChannelGroup` down. The API of `AsynchronousChannelGroup` is such that it considers its `ExecutorService` _"used exclusively"_ by the group<sup>1</sup>, and shuts it together with the group. This is achievable by decorating the shutdown-related methods of the the application-supplied executor. Ideally, the latter should be done with maintaining the contract of the methods (as far as I can see, it's doable). However, for simplicity, I think we could violate the contract by leaving `AsynchronousChannelGroup`-submitted tasks in the `ExecutorService` until they are executed - the resulting `Future`s will just be GCed.

<sup>1</sup> This can't be a hard requirement to not use the `ExecutorService` for anything else, because `AsynchronousChannelGroup` technically cannot observe whether the `ExecutorService` is used exclusively or not. And the "direct handoff or unbounded queuing" requirements guarantee that a task being submitted can't ever be rejected because there are either not enough threads to execute it, or the queue is full.

Reply:
Implemented ownership of executor service (closed by the driver, intended to be used exclusively by the driver). If a non-owned alternative is needed, we can add this in a future PR.